TFNetworkRecLayer.py

```diff
@@ -5293,56 +5293,73 @@ def post_process_final_rec_vars_outputs(self, rec_vars_outputs, seq_len):
+    if self.output.time_dim_axis:
+      assert self.output.time_dim_axis == source.time_dim_axis
+    if offset:  # Assumptions on the shape of `offset`.
+      assert offset.output.batch_dim_axis == 0, "Expect batch_dim_axis to be 0 for offset."
+      if offset.output.time_dim_axis is not None:
+        assert offset.output.time_dim_axis == 1, "Expect time_dim_axis to be 1 for offset."
     if add_to_input:
       assert source.dim == self.output.dim
     from TFUtil import get_positional_encoding
     if source.have_time_axis():
       length = tf.shape(source.placeholder)[source.time_dim_axis]
       if constant > -1:
-        position = constant * tf.ones([length], tf.int32)
-        signal = get_positional_encoding(num_channels=self.output.dim, position=position)  # (len,n_out)
+        position = constant * tf.ones([1, length], tf.int32)  # (1, len).
+        if offset:
+          position += offset.output.placeholder  # (batch, len).
+        signal = get_positional_encoding(num_channels=self.output.dim, position=position)  # (1, len) or (batch, len).
       else:
-        signal = get_positional_encoding(num_channels=self.output.dim, length=length)  # (len,n_out)
+        position = tf.range(length)  # (len,).
+        position = tf.expand_dims(position, axis=0)  # (1, len).
+        if offset:
+          position += offset.output.placeholder
+        signal = get_positional_encoding(num_channels=self.output.dim, position=position)
+      # signal has shape (1,len,n_out) or (batch,len,n_out).
+      if source.batch_dim_axis == 1:
+        assert source.time_dim_axis == 0,\
+          "Unexpected input form: batch_dim_axis 1 and time_dim_axis %s." % source.time_dim_axis
+        signal = tf.transpose(signal, [1, 0, 2])
     else:
       if constant > -1:
         position = tf.convert_to_tensor([constant])
       else:
         position = tf.convert_to_tensor([self.network.get_rec_step_index()])
-      signal = get_positional_encoding(num_channels=self.output.dim, position=position)  # (1,n_out)
-      signal = tf.squeeze(signal, axis=0)  # (n_out,)
-    if self.output.batch_dim_axis is not None:
-      signal = tf.expand_dims(signal, axis=self.output.batch_dim_axis)  # e.g. (len,batch,n_out)
+      if offset:
+        position += offset.output.placeholder  # (batch,).
+      signal = get_positional_encoding(num_channels=self.output.dim, position=position)  # (1,n_out) or (batch,n_out).
+      if source.batch_dim_axis != 0:
+        assert source.batch_dim_axis == 1 and source.feature_dim_axis == 0, "Unexpected input form."
+        signal = tf.transpose(signal, [1, 0])
+
     if add_to_input:
       signal += source.placeholder
-    else:
-      # tile to batch dimension explicitly, as batch_dim=1 will not be automatically unbroadcasted
-      tiles = [1] * self.output.batch_ndim
-      tiles[self.output.batch_dim_axis] = tf.shape(source.placeholder)[source.batch_dim_axis]
-      signal = tf.tile(signal, tiles)
```

Comment:
Why did you remove this? I think this is still needed (for `not add_to_input`).

Reply:
Putting this back.
In one of the previous versions, I was handling this upper in the code so I had removed this at some point.

Reply:
But for the moment I would put:
`tiles[self.source.batch_dim_axis]` instead of `tiles[self.output.batch_dim_axis]` .
We have to first decide whether we want to leave the output shape flexible or make it the same as the input, right?